### PR TITLE
Make access_log in http context configurable

### LINKED ISSUE #5625: Make access_log for main configurable
**Is your feature request related to a problem? Please describe.**
Currently it is not possible to globally configure log forward of access logs in a simple way. Logging high volume access logs to stdout means that they get forwarded to the log stack in Kubernetes. This can be problematic depending on the stack.

**Describe the solution you'd like**
Add a directive to the global Config map  where the main access_log directive can be overridden 

**Describe alternatives you've considered**
An alternative solution is to use a custom main template which is far from simple.

**Additional context**

### PR DESCRIPTION
**Pull request by hafe**
Closes #5625

### Proposed changes
See https://github.com/nginxinc/kubernetes-ingress/issues/5625

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [X] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have checked that all unit tests pass after adding my changes
- [X] I have updated necessary documentation
- [X] I have rebased my branch onto main
- [X] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### CODE REVIEW

**Review by ADubhlaoich — APPROVED** [review]
Doc change LGTM!

**Comment by jjngx** [comment]
Hi @hafe , could you please update the conflicting files?

**Comment by jjngx** [comment]
@hafe , could you please `gofumpt` files that are failing validation

File: docs/content/configuration/global-configuration/configmap-resource.md [hunk 1]
```diff
@@ -113,6 +113,7 @@ For more information, view the [VirtualServer and VirtualServerRoute resources](
 |ConfigMap Key | Description | Default | Example |
 | ---| ---| ---| --- |
 |*error-log-level* | Sets the global [error log level](https://nginx.org/en/docs/ngx_core_module.html#error_log) for NGINX. | *notice* |  |
+|*access-log* | Sets the directive [access log](https://nginx.org/en/docs/http/ngx_http_log_module.html#access_log). Only syslog is allowed. | ``/dev/stdout main`` | ``syslog:server=localhost:514`` |
```
Comment (j1m-ryan) [inline docs/content/configuration/global-configuration/configmap-resource.md:116]: > Only syslog is allowed

Or standard out also

**Comment by j1m-ryan** [comment]
Thanks for the pr @hafe, @vepatel is looking at this pr now.

File: docs/content/configuration/global-configuration/configmap-resource.md (hunk 1 above)
Comment (hafe) [reply docs/content/configuration/global-configuration/configmap-resource.md:116]: You mean documented? Stdout is the default as it is now

File: internal/configs/version1/config.go [hunk 2]
```diff
@@ -191,6 +191,7 @@ type Location struct {
 
 // MainConfig describe the main NGINX configuration file.
 type MainConfig struct {
+	AccessLog                          string
 	AccessLogOff                       bool
```
Comment (vepatel) [inline internal/configs/version1/config.go:195]: @hafe is this value still being used anywhere?

File: internal/configs/version1/__snapshots__/template_test.snap [hunk 3]
```diff
@@ -34,7 +34,7 @@ http {
         default $upstream_trailer_grpc_status;
         '' $sent_http_grpc_status;
     }
-    access_log  /dev/stdout  main;
+    access_log ;
```
Comment (vepatel) [inline internal/configs/version1/__snapshots__/template_test.snap:37]: @hafe this seems a bit odd as only allowed configs as per this PR are:
`access_log off;`
`access_log /dev//stdout main`
`access_log syslog:xxxx`

File: docs/content/configuration/global-configuration/configmap-resource.md (hunk 1 above)
Comment (vepatel) [reply docs/content/configuration/global-configuration/configmap-resource.md:116]:
```suggestion
|*access-log* | Sets the directive [access log](https://nginx.org/en/docs/http/ngx_http_log_module.html#access_log). only allowed values are standard out (default) and a syslog destination. | ``/dev/stdout main`` | ``syslog:server=localhost:514`` |
```

File: internal/configs/version1/config.go (hunk 2 above)
Comment (hafe) [reply internal/configs/version1/config.go:195]: Yes can be removed.

File: internal/configs/version1/__snapshots__/template_test.snap (hunk 3 above)
Comment (hafe) [reply internal/configs/version1/__snapshots__/template_test.snap:37]: I don't know how it works. I just ran "make test-update-snaps" and ended up with this.

File: internal/configs/version1/config.go (hunk 2 above)
Comment (vepatel) [reply internal/configs/version1/config.go:195]: @hafe can you please remove this var?

File: internal/configs/version1/__snapshots__/template_test.snap (hunk 3 above)
Comment (vepatel) [reply internal/configs/version1/__snapshots__/template_test.snap:37]: @hafe can you please add 
`AccessLog:                      "/dev/stdout main",`
after https://github.com/hafe/kubernetes-ingress/blob/feat/access-log-5625/internal/configs/version1/template_test.go#L2041
and then run `make test-update-snaps`. it should generate the correct config

File: internal/configs/configmaps.go [hunk 4]
```diff
@@ -207,11 +207,21 @@ func ParseConfigMap(cfgm *v1.ConfigMap, nginxPlus bool, hasAppProtect bool, hasA
 		cfgParams.MainErrorLogLevel = errorLogLevel
 	}
 
+	if accessLog, exists := cfgm.Data["access-log"]; exists {
+		if !strings.HasPrefix(accessLog, "syslog:") {
```
Comment (vepatel) [inline internal/configs/configmaps.go:211]: seems to allow stuff other than syslog
```
Name:         test-release-nginx-ingress
Namespace:    default
Labels:       app.kubernetes.io/instance=test-release
              app.kubernetes.io/managed-by=Helm
              app.kubernetes.io/name=nginx-ingress
              app.kubernetes.io/version=3.7.0
              helm.sh/chart=nginx-ingress-1.4.0
Annotations:  meta.helm.sh/release-name: test-release
              meta.helm.sh/release-namespace: default

Data
====
access-log:
----
yslog:server=syslog-svc.default:514

BinaryData
====

Events:
  Type    Reason   Age                From                      Message
  ----    ------   ----               ----                      -------
  Normal  Updated  11s (x5 over 21m)  nginx-ingress-controller  Configuration from default/test-release-nginx-ingress was updated
```

Comment (hafe) [reply internal/configs/configmaps.go:211]: OK any idea why?

Comment (hafe) [reply internal/configs/configmaps.go:211]: OK it's allowed but ignored (with an error log) which is the case for all other configmap items as far as I understand

Comment (vepatel) [reply internal/configs/configmaps.go:211]: @hafe we would ideally like it to make configmap invalid

**Comment by jjngx** [comment]
@hafe could you please update the branch? I am reviewing changes atm.

File: internal/configs/configmaps.go (hunk 4 above)
Comment (jjngx) [reply internal/configs/configmaps.go:211]: looking into this case atm

**Comment by hafe** [comment]
> @hafe could you please update the branch? I am reviewing changes atm.

You mean just rebase it from main?

**Comment by jjngx** [comment]
> > @hafe could you please update the branch? I am reviewing changes atm.
> 
> You mean just rebase it from main?

Yes, please rebase.

File: internal/configs/configmaps.go (hunk 4 above)
Comment (jjngx) [reply internal/configs/configmaps.go:211]: This works as expected (implemented in the code atm).

When user sets the access-log-off to `true` and the value of `access-log` does not meet the condition (starts with `syslog`) the log is set to it's default value: `access_log /dev/stdout main;`

**Comment by hafe** [comment]
I can do that. It should really be a warning log also

**Comment by vepatel** [comment]
@jjngx @hafe lets add it to the docs as well for clarity and since we can't report it through configmap event:

> ` access_log` value will be set to `/dev/stdout main` in-case user tries to configure it with location other than a syslog.

**Comment by jjngx** [comment]
@hafe could you please also update template tests?

**Comment by hafe** [comment]
Not sure what you mean by that

**Comment by jjngx** [comment]
> Not sure what you mean by that

could you run `make test` from the root dir of the project and check if the `internal/configs/version1/__snapshots__/template_test.snap` changes? When you look in the file there are new lines: `access_log ;` that indicates invalid (from NGINX perspective) lines.

**Comment by hafe** [comment]
No nothing changes.

**Comment by hafe** [comment]
Seems to be OK now after latest updates. Thanks for the help!

**Review by jjngx — APPROVED** [review]
👍🏻

**Review by j1m-ryan — APPROVED** [review]

**Pull request merged**